skToInflux.js

```diff
@@ -15,87 +15,148 @@
+  }
+  return fields
+}
 
 module.exports = {
   deltaToPointsConverter: (
     selfContext,
     recordTrack,
     shouldStore,
     resolution,
+    storeOtherVessels,
     useDeltaTimestamp = false
   ) => {
     return delta => {
+    
       if (delta.context === 'vessels.self') {
-        delta.context = selfContext
+        delta.context = app.selfId
       }
       let points = []
-      if (delta.updates && delta.context === selfContext) {
+      if (delta.updates && (storeOtherVessels || delta.context === selfContext)) {
         delta.updates.forEach(update => {
           if (update.values) {
             let date = new Date(update.timestamp)
             let time = date.getTime()
+            let tags = { context: delta.context }
             points = update.values.reduce((acc, pathValue) => {
               if (pathValue.path === 'navigation.position') {
                 if (
-                  recordTrack &&
-                  time - lastPositionStored > 1000
+                  recordTrack
+                    && (!lastPositionStored[delta.context]
+                        || time - lastPositionStored[delta.context] > 1000)
                 ) {
-                  acc.push({
+                  const point = {
                     measurement: pathValue.path,
-                    fields: {
-                      value: JSON.stringify([
-                        pathValue.value.longitude,
-                        pathValue.value.latitude
-                      ])
-                    }
-                  })
-                  lastPositionStored = Date.now()
+                    tags: tags,
+                    fields: addSource(update, {
+                      jsonValue: JSON.stringify({
+                        longitude: pathValue.value.longitude,
+                        latitude: pathValue.value.latitude
+                      })
+                    })
+                  }
+                  if (useDeltaTimestamp) {
+                    point.timestamp = date
+                  }
+                  acc.push(point)
+                  lastPositionStored[delta.context] = time
                 }
               } else if (shouldStore(pathValue.path)) {
-                if ( !lastUpdates[pathValue.path] || time - lastUpdates[pathValue.path] > resolution ) {
-                  lastUpdates[pathValue.path] = time
+                if ( !lastUpdates[delta.context] || !lastUpdates[delta.context][pathValue.path] || time - lastUpdates[delta.context][pathValue.path] > resolution ) {
+                  if ( !lastUpdates[delta.context] )
+                    lastUpdates[delta.context] = {}
+                  lastUpdates[delta.context][pathValue.path] = time
                   if (pathValue.path === 'navigation.attitude') {
                     if (typeof pathValue.value.pitch === 'number' && !isNaN(pathValue.value.pitch)) {
-                      acc.push({
+                      const point = {
```

Comment:
I know the copypaste for pitch, yaw and roll was already there, but maybe we should be boyscouts and fix this with a single `handleAttitude` call and there iterate over `['pitch', 'yaw','roll']`?

I have to say I am pretty unhappy about this schema aware, special code for attitude in general. Could this be generalised so that it works for any object-valued pathValue?

Reply:
I extracted that out into a function and did the iteration. I'd like to tackle the second part in another PR